vulnerabilities/importers/mozilla.py

```diff
@@ -0,0 +1,180 @@
+
+        advisories.append(
+            Advisory(
+                summary=summary,
+                vulnerability_id=cve if is_cve(cve) else "",
+                impacted_package_urls=[],
+                resolved_package_urls=fixed_package_urls,
+                references=references,
+            )
+        )
+
+    return advisories
+
+
+def get_advisories_from_md(mfsa_id, lines) -> List[Advisory]:
+    yamltext, mdtext = split_markdown_front_matter(lines.read())
+    data = yaml.safe_load(yamltext)
+    data["mfsa_id"] = mfsa_id
+
+    fixed_package_urls = get_package_urls(data.get("fixed_in"))
+    references = get_yml_references(data)
+    cves = re.findall(r"CVE-\d+-\d+", yamltext + mdtext, re.IGNORECASE)
+    for cve in cves:
+        references.append(
+            Reference(
+                reference_id=cve,
+                url=f"https://cve.mitre.org/cgi-bin/cvename.cgi?name={cve}",
+            )
+        )
+
+    description = html_get_p_under_h3(markdown(mdtext), "description")
+
+    return [
+        Advisory(
+            summary=description,
+            vulnerability_id="",
+            impacted_package_urls=[],
+            resolved_package_urls=fixed_package_urls,
+            references=references,
+        )
+    ]
+
+
+def html_get_p_under_h3(html, h3: str):
+    soup = BeautifulSoup(html, features="lxml")
+    h3tag = soup.find("h3", text=lambda txt: txt.lower() == h3)
+    p = ""
+    if h3tag:
+        for tag in h3tag.next_siblings:
+            if tag.name:
+                if tag.name != "p":
+                    break
+                p += tag.get_text()
+    return p
+
+
+def mfsa_id_from_filename(filename):
+    match = MFSA_FILENAME_RE.search(filename)
+    if match:
+        return "mfsa" + match.group(1)
+
+    return None
+
+
+def get_package_urls(pkgs: List[str]) -> List[PackageURL]:
+    package_urls = [
+        PackageURL(
+            type="mozilla",
+            # pkg is of the form "Firefox ESR 1.21" or "Thunderbird 2.21"
+            name=pkg.rsplit(None, 1)[0],
+            version=pkg.rsplit(None, 1)[1],
+        )
+        for pkg in pkgs
+        if pkg
+    ]
+    return package_urls
+
+
+def get_yml_references(data: any) -> List[Reference]:
+    """
+    Returns a list of references
+    Currently only considers the given mfsa as a reference
+    """
+    # FIXME: Needs improvement
+    # Should we add 'bugs' section in references too?
+    # Should we add 'impact'/severity of CVE in references too?
+    # If yes, then fix alpine_linux importer as well
+    # Otherwise, do we need severity field for adversary as well?
+
+    severities = ["critical", "high", "medium", "low", "none"]
+    severity = "none"
+    if data.get("impact"):
+        impact = data.get("impact").lower()
+        for s in severities:
+            if s in impact:
+                severity = s
+                break
+
+    return [
+        Reference(
+            reference_id=data["mfsa_id"],
```

Comment:
Create a variable for `data["mfsa_id"]` and reuse it. Do not get multiple times.

Reply:
I added the `mfsa_id` in `data` to make sure everything we need about the advisory is at one place, inside `data`. 
https://github.com/nexB/vulnerablecode/blob/5997b26ba35d3cff83c07ee6f6fd2c81b4f526ae/vulnerabilities/importers/mozilla.py#L65

Is there something wrong with this approach?

Reply:
Why pass down a mapping when you can pass named arguments? 
It is much better to give names.
Something such as:
```Python
def get_reference(mfsa_id, impact):
    """
    Return a reference for tan mfsa and is impact value.
    """
    known_impacts = set(["critical", "high", "medium", "low", "none"])

    impact = impact or ''
    impact = impact.lower()
    severity = known_impacts.get(impact)

    severities=[]
    if severity:
        severities = [VulnerabilitySeverity(scoring_systems["generic_textual"], severity)]

    return Reference(
        reference_id=mfsa_id,
        url=f"https://www.mozilla.org/en-US/security/advisories/{mfsa_id}",
        severities=severities,
    )

```

Also do not return a list when you return a single value. I appreciate planning for the future, but that's usually not needed. We can always refactor later if needed.


And add other impacts that may be funky to `known_impacts` that you can turn in a mapping.
I would also raise an exception or some warning if there is an impact that is not known and is not in the mapping.

Reply:
Thanks.
I've modified it to look like this
```py
def get_yml_references(mfsa_id, impact, advisories) -> List[Reference]:
    """
    Return a list of references
    Considers the following as references:
        - mfsa
        - cve
        - bugzilla or direct reference
    """

    known_impacts = { 
            "severe": "critical",
            "critical": "critical",
            "critical (local)": "critical",
            "high": "high",
            "moderate": "moderate",
            "medium": "moderate",
            "moderate (on a multiuser computer)": "moderate",
            "low": "low",
            "critical (firefox 2.0 not affected in default configuration)": "critical",
            "low (critical in gecko 1.9.1 and earlier)": "low",
            "moderate to critical": "moderate",
            }


    impact = impact or ''
    impact = impact.lower()
    severity = known_impacts.get(impact)
    if severity:
        severities=[VulnerabilitySeverity(scoring_systems["generic_textual"], severity)]
    else:
        warnings.warn(f'{mfsa_id}: unknown severity "{severity}"')

    # Add mfsa as reference
    references = [
            Reference(
                reference_id=data["mfsa_id"],
                url=f"https://www.mozilla.org/en-US/security/advisories/{data['mfsa_id']}",
                severities=severities,
                )
            ]

    advisories = data.get("advisories", {})
    for advisory in advisories:
        # Add advisory cve as reference
    ...
    ...
```
I would be needing the references list as now I've also added the bugs and cve itself as a reference.

Reply:
Have you pushed these updates?